ext/compressors/lz4/lz4_compress.c

```diff
@@ -277,8 +277,9 @@ lz4_compress_raw(WT_COMPRESSOR *compressor, WT_SESSION *session,
 	(void)final;
 
 	sourceSize = (int)offsets[slots];		/* Type conversion */
-	targetDestSize =
-	    (int)((dst_len < page_max ? dst_len : page_max) - extra);
+	targetDestSize = (int)
+	    (((dst_len - sizeof(LZ4_PREFIX)) < page_max ?
+	    (dst_len - sizeof(LZ4_PREFIX)) : page_max) - extra);
```

Comment:
I think it's a bug in the`zlib_compress_raw()` function.

In short, `page_max` is the target page size taken from `WT_BTREE.{maxleafpage,maxintlpage}`, and `dst_len` is either the return value of the `WT_COMPRESSOR.pre_size` function or the source size. Currently, `page_max` can be larger or smaller than `dst_len`.

Since zlib doesn't have a pre_size function, `dst_len` is the source size, and the source size may be smaller than the target page size, which `zlib_compress_raw()` doesn't take into account. We don't fail because zlib is very, very unlikely to exceed the source size during compression, but I think we'd drop core if it ever did grow.

I've pushed a change in 78f145c to fix the zlib code, explain what's going on and to make the lz4 and zlib code look the same.

We could also change the reconciliation code to ensure the destination buffer is never smaller than the target page size (currently we only guarantee it's never smaller than the return of the pre_size function or the length of the source if there's no pre_size function). That might be safer for future raw compression functions.